### PR TITLE
Implement reading database data in SILO API

### LINKED ISSUE #16: Implement reading database data in SILO API
As a first step: When starting the API, read the data, process it, and load it into the database. This has to be split up in a further step into a preprocessing pipeline and reading a serialized state.

When the relevant steps are migrated, delete the current [`main.cpp`](https://github.com/GenSpectrum/LAPIS-SILO/blob/6908c7f3a6ac85008af3606d0222ea58759ce954/src/main.cpp) and only leave the api main method. (Make sure to remove the dependency on readline)

### PR DESCRIPTION
**Pull request by JonasKellerer**
Resolves #16

### CODE REVIEW

**Comment by chaoran-chen** [comment]
The sequence files are quite big, so I wonder whether we really want to include them in this repository. They are not huge but still several MBs large and if we update the test data in the future, this can bloat up the repository by quite a bit. Should we maybe store them in a separate repository instead?

File: src/database.cpp [hunk 1]
```diff
@@ -4,14 +4,69 @@
...
 #include <tbb/blocked_range.h>
 #include <tbb/parallel_for_each.h>
 
+const std::string REFERENCE_GENOME_FILENAME = "reference_genome.txt";
+const std::string PANGO_ALIAS_FILENAME = "pango_alias.txt";
+
+std::vector<std::string> silo::Database::initGlobalReference(const std::string& working_directory) {
+   std::ifstream reference_file(working_directory + REFERENCE_GENOME_FILENAME);
+   if (!reference_file) {
+      std::cerr << "Expected file " << working_directory << REFERENCE_GENOME_FILENAME << "." << std::endl;
```
Comment (fengelniederhammer) [inline src/database.cpp:24]: Let's get rid of the `std::cerr` here and in `initAliasKey`, while we're on it. Also let's improve the exception messages a bit to disclose more details. I imagine something like "Global reference genome file $file does not exist".

File: src/database.cpp [hunk 2]
```diff
@@ -163,23 +217,6 @@ void silo::Database::flipBitmaps() {
    });
 }
 
-/*
-void silo::Database::analyse() {
```
Comment (fengelniederhammer) [inline src/database.cpp:167]: :+1:

File: src/database.cpp [hunk 3]
```diff
@@ -567,3 +605,52 @@ void silo::Database::load(const std::string& save_dir) {
       ia >> partitions[i];
    });
 }
+void silo::Database::preprocessing(
+   const PreprocessingConfig& config) {
```
Comment (fengelniederhammer) [inline src/database.cpp:609]: I guess this method header fits in a single line

File: src/database.cpp [hunk 4]
```diff
@@ -567,3 +605,52 @@ void silo::Database::load(const std::string& save_dir) {
...
+   std::cout << "build_pango_defs" << std::endl;
+   std::ifstream metadata_stream(config.metadata_file.relative_path());
+   pango_def = std::make_unique<pango_descriptor_t>(silo::build_pango_defs(alias_key, metadata_stream));
+
+   std::cout << "build_partitioning_descriptor" << std::endl;
+   part_def = std::make_unique<partitioning_descriptor_t>(silo::build_partitioning_descriptor(*pango_def, architecture_type::max_partitions));
```
Comment (fengelniederhammer) [inline src/database.cpp:615]:
```suggestion
   partitioning_descriptor = std::make_unique<partitioning_descriptor_t>(silo::build_partitioning_descriptor(*pango_def, architecture_type::max_partitions));
```

File: src/database.cpp [hunk 5]
```diff
@@ -567,3 +605,52 @@ void silo::Database::load(const std::string& save_dir) {
       ia >> partitions[i];
    });
 }
+void silo::Database::preprocessing(
+   const PreprocessingConfig& config) {
+   std::cout << "build_pango_defs" << std::endl;
+   std::ifstream metadata_stream(config.metadata_file.relative_path());
+   pango_def = std::make_unique<pango_descriptor_t>(silo::build_pango_defs(alias_key, metadata_stream));
```
Comment (fengelniederhammer) [inline src/database.cpp:612]:
```suggestion
   pango_descriptor = std::make_unique<pango_descriptor_t>(silo::build_pango_defs(alias_key, metadata_stream));
```

File: src/database.cpp [hunk 6]
```diff
@@ -567,3 +605,52 @@ void silo::Database::load(const std::string& save_dir) {
...
+
+   std::cout << "Dictionary" << std::endl;
+   dict = std::make_unique<Dictionary>();
+   for (size_t i = 0; i < part_def->partitions.size(); ++i) {
+      const auto& part = part_def->partitions.at(i);
+      for (unsigned j = 0; j < part.chunks.size(); ++j) {
+         std::string name = std::string(config.partition_folder.relative_path()) + chunk_string(i, j) + std::string(config.metadata_file.extension());
+         std::ifstream meta_in(name);
+         if (!meta_in) {
+            std::cerr << "Meta_data file " << name << " not found." << std::endl;
```
Comment (fengelniederhammer) [inline src/database.cpp:644]:
```suggestion
            throw silo::PreprocessingException("Metadata partition file " + name + " not found.");
```
I think a dedicated `PreprocessingException` makes sense to have more specialized exceptions.

File: src/database.test.cpp [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#include "gtest/gtest.h"
+#include <iostream>
```
Comment (fengelniederhammer) [inline src/database.test.cpp:2]: CLion complains that this include is unused

File: src/database.test.cpp [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+#include <silo/preprocessing_config.h>
+
+TEST(database_test, should_build_database_without_errors) {
+
+   const std::string input_directory("./testBaseData/");
+   const std::string output_directory("./build/");
+   auto config = silo::PreprocessingConfig(input_directory, output_directory, "minimal_set.tsv", "minimal_set.fasta");
+   auto database = std::make_shared<silo::Database>(input_directory);
+
+   EXPECT_NO_THROW(database->preprocessing(config));
```
Comment (fengelniederhammer) [inline src/database.test.cpp:14]: We should instead check that the database is not empty. Spontaneous idea: `silo::Database::get_db_info()` should do the job.

File: src/main.cpp [hunk 9]
```diff
@@ -180,7 +180,7 @@ int handle_command(Database& db, std::vector<std::string> args) {
          return 0;
       }
       std::cout << "Build pango_def from file " << meta_input_str << std::endl;
-      db.pango_def = std::make_unique<pango_descriptor_t>(silo::build_pango_defs(db.get_alias_key(), meta_input));
+      db.pango_def = std::make_unique<pango_descriptor_t>(silo::build_pango_defs(db.getAliasKey(), meta_input));
```
Comment (fengelniederhammer) [inline src/main.cpp:183]: Maybe simply delete the commands that we already migrated?

File: include/local.cmake [hunk 10]
```diff
@@ -20,4 +20,5 @@ set(
         include/silo/database.h
         include/silo/benchmark.h
         include/silo/bootstrap.h
+        include/silo/preprocessing_config.h
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply include/local.cmake:23]: Not needed because already include ${CMAKE_SOURCE_DIR}/include in CMakeLists.txt

**Review by fengelniederhammer — APPROVED** [review]

**Pull request merged**